irc/websocket.go

```diff
@@ -0,0 +1,70 @@
+package irc
+
+import (
+	"bytes"
+	"errors"
+	"github.com/gorilla/websocket"
+	"net/http"
+	"time"
+	"unicode/utf8"
+)
+
+var wsUpgrader = websocket.Upgrader{
+	ReadBufferSize:  2 * 1024,
+	WriteBufferSize: 2 * 1024,
+	// If a WS session contains sensitive information, and you choose to use
+	// cookies for authentication (during the HTTP(S) upgrade request), then
+	// you should check that Origin is a domain under your control. If it
+	// isn't, then it is possible for users of your site, visiting a naughty
+	// Origin, to have a WS opened using their credentials. See
+	// http://www.christian-schneider.net/CrossSiteWebSocketHijacking.html#main.
+	// We don't care about Origin because the (IRC) authentication is contained
+	// in the WS stream -- the WS session is not privileged when it is opened.
+	CheckOrigin: func(r *http.Request) bool { return true },
+}
+
+// WSContainer wraps a WebSocket connection so that it implements net.Conn
+// entirely.
+type WSContainer struct {
+	*websocket.Conn
+}
+
+func (ws WSContainer) Read(b []byte) (n int, err error) {
+	var messageType int
+	var bytes []byte
+
+	for {
+		messageType, bytes, err = ws.ReadMessage()
+		if messageType == websocket.TextMessage {
+			n = copy(b, bytes)
+			return
+		}
+		if len(bytes) == 0 {
+			return 0, nil
+		}
+		// Throw other kind of messages away.
+	}
+	// We don't want to return (0, nil) here because that would mean the
+	// connection is closed (Read calls must block until data is received).
+}
+
+func (ws WSContainer) Write(b []byte) (n int, err error) {
+	if !utf8.Valid(b) {
+		return 0, errors.New("outgoing WebSocket message isn't valid UTF-8")
+	}
+
+	b = bytes.TrimSuffix(b, []byte("\r\n"))
+	n = len(b)
```

Comment:
I believe the underlying API being implemented here is [io.Writer](https://golang.org/pkg/io/#Writer), which says "Write must return a non-nil error if it returns n < len(p)."

I think it's correct to report either that all the bytes of the original message (including the last 2) were written and there was no error, or else report an error condition. Otherwise something might hypothetically retry the write.

Now that I think about it, how is it that `WSContainer` implements all of the `net.Conn` interface? Does it inherit methods like `LocalAddr` and `RemoteAddr` from `websocket.Conn`?

Reply:
> I believe the underlying API being implemented here is [io.Writer](https://golang.org/pkg/io/#Writer), which says "Write must return a non-nil error if it returns n < len(p)."
> 
> I think it's correct to report either that all the bytes of the original message (including the last 2) were written and there was no error, or else report an error condition. Otherwise something might hypothetically retry the write.

Good point, let's just return `len(b), nil` when the write is successful.

> Now that I think about it, how is it that `WSContainer` implements all of the `net.Conn` interface? Does it inherit methods like `LocalAddr` and `RemoteAddr` from `websocket.Conn`?

Yup, `websocket.Conn` just misses those three methods in `irc/websocket.go` to implement the full `net.Conn` interface.

https://godoc.org/github.com/gorilla/websocket#Conn.Close
https://godoc.org/github.com/gorilla/websocket#Conn.LocalAddr
https://godoc.org/github.com/gorilla/websocket#Conn.RemoteAddr
https://godoc.org/github.com/gorilla/websocket#Conn.SetReadDeadline
https://godoc.org/github.com/gorilla/websocket#Conn.SetWriteDeadline
https://golang.org/pkg/net/#Conn